File: src/unicode_python/security/covert/variation_selector_payload.py
"""Detection of GlassWorm-class invisible payloads encoded in
Unicode variation selectors.

Threat model.  Tier A1.  Adversary crafts an input consisting of
one visible base codepoint followed by a sequence of variation-
selector codepoints (U+FE00..U+FE0F union U+E0100..U+E01EF) that
the receiving renderer treats as a no-op glyph variant but that
a downstream string-processing layer (e.g. an LLM tokenizer or a
clipboard pipeline) preserves byte-for-byte.  Decoding pairs of
VS codepoints back into bytes recovers an arbitrary payload.

This port treats every variation-selector occurrence after a base
codepoint as suspicious.  The Lean reference additionally exempts
(base, VS) pairs that appear in StandardizedVariants.txt and
emoji-presentation pairs — those exemptions require UCD tables.
The coarser port is more conservative: legitimate math / Mongolian
/ Egyptian / CJK Compat / emoji-presentation VS uses will be
flagged.
"""

from dataclasses import dataclass, field
from pathlib import Path
from typing import Union

from ..calculus import ClassificationKind

# ──────────────────────────────────────────────────────────────────────
# Authoritative legal (base, VS) pair set — UCD StandardizedVariants
# + UTS #51 emoji-variation-sequences.  Loaded lazily on first use.
# ──────────────────────────────────────────────────────────────────────

_DATA_DIR = Path(__file__).resolve().parent.parent.parent / "data"


def _parse_legal_variation_pairs() -> set[tuple[int, int]]:
    out: set[tuple[int, int]] = set()
    for fname in (
        "StandardizedVariants.txt",
        "emoji-variation-sequences.txt",
    ):
        path = _DATA_DIR / fname
        with path.open("r", encoding="utf-8") as f:
            for raw_line in f:
                # Strip comment then take up to first ';'.
                hash_idx = raw_line.find("#")
                body = raw_line if hash_idx < 0 else raw_line[:hash_idx]
                semi_idx = body.find(";")
                pair_part = body if semi_idx < 0 else body[:semi_idx]
                tokens = pair_part.split()
                if len(tokens) < 2:
                    continue
                try:
                    base = int(tokens[0], 16)
                    vs = int(tokens[1], 16)
                except ValueError:
                    continue
                out.add((base, vs))
    return out


_LEGAL_PAIRS: set[tuple[int, int]] | None = None


def is_registered_variation_pair(base: int, vs: int) -> bool:
    """True iff ``(base, vs)`` is a registered variation sequence
    per UCD StandardizedVariants.txt or UTS #51
    emoji-variation-sequences.txt.  Used to exempt legitimate
    math / emoji-presentation variants from IllegalTarget
    false-positives."""
    global _LEGAL_PAIRS
    if _LEGAL_PAIRS is None:
        _LEGAL_PAIRS = _parse_legal_variation_pairs()
    return (base, vs) in _LEGAL_PAIRS


def is_variation_selector(cp: int) -> bool:
    if 0xFE00 <= cp <= 0xFE0F:
        return True
    if 0xE0100 <= cp <= 0xE01EF:
        return True
    if 0x180B <= cp <= 0x180D:
        return True
    return False


def vs_to_nibble(cp: int) -> int | None:
    """Decode a single VS codepoint to its nibble value in [0, 255].
    Uses GlassWorm's bit layout: VS1..VS16 → nibbles 0..15,
    VS17..VS256 → nibbles 16..255.  Mongolian FVS codepoints
    (180B..180D) return ``None``."""
    if 0xFE00 <= cp <= 0xFE0F:
        return cp - 0xFE00
    if 0xE0100 <= cp <= 0xE01EF:
        return cp - 0xE0100 + 16
    return None


@dataclass(frozen=True, slots=True)
class DirectPayload:
    decoded: str


@dataclass(frozen=True, slots=True)
class IllegalTarget:
    target_cp: int
    vs_cp: int


@dataclass(frozen=True, slots=True)
class RepeatedBase:
    base_cp: int
    vs_count: int


SubThreat = Union[DirectPayload, IllegalTarget, RepeatedBase]


def sub_threat_tag(sub: SubThreat) -> str:
    if isinstance(sub, DirectPayload):
        return "DirectPayload"
    if isinstance(sub, IllegalTarget):
        return "IllegalTarget"
    return "RepeatedBase"


@dataclass(slots=True)
class Verdict:
    kind: ClassificationKind
    sub: SubThreat | None = None
    vs_positions: list[int] = field(default_factory=list)
    recovered_bytes: bytes = b""


def _decode_vs_run(input_cps: list[int], positions: list[int]) -> bytes:
    out = bytearray()
    high: int | None = None
    for p in positions:
        n = vs_to_nibble(input_cps[p])
        if n is None:
            continue
        if high is None:
            high = n
        else:
            out.append(((high << 4) | n) & 0xFF)
            high = None
    return bytes(out)


def _all_same_vs(input_cps: list[int], positions: list[int]) -> bool:
    if not positions:
        return True
    cp0 = input_cps[positions[0]]
    return all(input_cps[p] == cp0 for p in positions)


def _lossy_ascii(payload: bytes) -> str:
    out = []
    for b in payload:
        if (0x20 <= b <= 0x7E) or b in (0x09, 0x0A, 0x0D):
            out.append(chr(b))
        else:
            out.append("?")
    return "".join(out)


def detect(input_cps: list[int]) -> Verdict:
    v = Verdict(kind=ClassificationKind.CLEAR)
    v.vs_positions = [i for i, cp in enumerate(input_cps) if is_variation_selector(cp)]
    if not v.vs_positions:
        return v

    v.recovered_bytes = _decode_vs_run(input_cps, v.vs_positions)

    # Single-VS exemption: if exactly one VS follows a base AND the
    # (base, VS) pair is registered in StandardizedVariants or
    # emoji-variation-sequences, return Clear (legitimate variant).
    if len(v.vs_positions) == 1:
        p = v.vs_positions[0]
        if p > 0:
            base = input_cps[p - 1]
            vs = input_cps[p]
            if is_registered_variation_pair(base, vs):
                return v  # Clear — registered variant

    v.kind = ClassificationKind.HAZARD

    if len(v.vs_positions) >= 4 and _all_same_vs(input_cps, v.vs_positions):
        p0 = v.vs_positions[0]
        base = 0 if p0 == 0 else input_cps[p0 - 1]
        v.sub = RepeatedBase(base_cp=base, vs_count=len(v.vs_positions))
    elif v.recovered_bytes:
        v.sub = DirectPayload(decoded=_lossy_ascii(v.recovered_bytes))
    else:
        p = v.vs_positions[0]
        target = 0 if p == 0 else input_cps[p - 1]
        v.sub = IllegalTarget(target_cp=target, vs_cp=input_cps[p])
    return v


__all__ = [
    "DirectPayload",
    "IllegalTarget",
    "RepeatedBase",
    "SubThreat",
    "Verdict",
    "detect",
    "is_variation_selector",
    "sub_threat_tag",
    "vs_to_nibble",
]

File: src/security/covert/variation_selector_payload.rs
//! Detection of GlassWorm-class invisible payloads encoded in
//! Unicode variation selectors.
//!
//! Threat model.  Tier A1.  Adversary crafts an input consisting of
//! one visible base codepoint followed by a sequence of variation-
//! selector codepoints (U+FE00..U+FE0F union U+E0100..U+E01EF) that
//! the receiving renderer treats as a no-op glyph variant but that
//! a downstream string-processing layer (e.g. an LLM tokenizer or a
//! clipboard pipeline) preserves byte-for-byte.  Decoding pairs of
//! VS codepoints back into bytes recovers an arbitrary payload.
//!
//! Exempts (base, VS) pairs that appear in
//! `StandardizedVariants.txt` (1127 registered variation
//! sequences) and `emoji-variation-sequences.txt` (371 emoji
//! presentation pairs) per UCD 17.0 / UTS #51.  Without this
//! exemption the detector false-positives on legitimate East-
//! Asian text containing CJK Compatibility Ideograph + VS pairs
//! and registered math-symbol variants — a state-level credibility
//! issue for banking / govt deployments processing CJK identifiers.

use std::collections::HashSet;
use std::sync::OnceLock;

use crate::security::ClassificationKind;

// ──────────────────────────────────────────────────────────────────────
// Authoritative legal (base, VS) pair set
// ──────────────────────────────────────────────────────────────────────

const STANDARDIZED_VARIANTS_RAW: &str =
    include_str!("../../../data/StandardizedVariants.txt");
const EMOJI_VARIATION_SEQUENCES_RAW: &str =
    include_str!("../../../data/emoji-variation-sequences.txt");

fn parse_hex_u32(s: &str) -> Option<u32> {
    u32::from_str_radix(s.trim(), 16).ok()
}

fn parse_legal_pairs() -> HashSet<(u32, u32)> {
    let mut out = HashSet::new();
    for source in [STANDARDIZED_VARIANTS_RAW, EMOJI_VARIATION_SEQUENCES_RAW] {
        for raw_line in source.lines() {
            let body = match raw_line.find('#') {
                Some(idx) => &raw_line[..idx],
                None => raw_line,
            };
            let stripped = body.trim();
            if stripped.is_empty() {
                continue;
            }
            // Format: "<base-hex> <vs-hex>; <description>" — we only
            // need the first two whitespace-separated hex tokens
            // before the ';'.
            let semi_idx = stripped.find(';').unwrap_or(stripped.len());
            let pair_part = &stripped[..semi_idx];
            let mut tokens = pair_part.split_whitespace();
            let (Some(base_str), Some(vs_str)) = (tokens.next(), tokens.next())
            else {
                continue;
            };
            let (Some(base), Some(vs)) =
                (parse_hex_u32(base_str), parse_hex_u32(vs_str))
            else {
                continue;
            };
            out.insert((base, vs));
        }
    }
    out
}

/// True iff `(base, vs)` is a registered variation sequence per
/// UCD 17.0 StandardizedVariants.txt or UTS #51
/// emoji-variation-sequences.txt.  Used to exempt legitimate
/// CJK Compatibility / math / emoji-presentation variants from
/// IllegalTarget false-positives.
pub fn is_registered_variation_pair(base: u32, vs: u32) -> bool {
    static SET: OnceLock<HashSet<(u32, u32)>> = OnceLock::new();
    SET.get_or_init(parse_legal_pairs).contains(&(base, vs))
}

pub fn is_variation_selector(cp: u32) -> bool {
    matches!(cp, 0xFE00..=0xFE0F | 0xE0100..=0xE01EF | 0x180B..=0x180D)
}

/// Decode a single VS codepoint to its nibble value in [0, 255].
/// Uses GlassWorm's bit layout: VS1..VS16 → nibbles 0..15,
/// VS17..VS256 → nibbles 16..255.  Mongolian FVS codepoints
/// (180B..180D) return `None`.
pub fn vs_to_nibble(cp: u32) -> Option<u32> {
    if (0xFE00..=0xFE0F).contains(&cp) {
        Some(cp - 0xFE00)
    } else if (0xE0100..=0xE01EF).contains(&cp) {
        Some(cp - 0xE0100 + 16)
    } else {
        None
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum SubThreat {
    DirectPayload { decoded: String },
    IllegalTarget { target_cp: u32, vs_cp: u32 },
    RepeatedBase { base_cp: u32, vs_count: usize },
}

impl SubThreat {
    pub fn tag(&self) -> &'static str {
        match self {
            SubThreat::DirectPayload { .. } => "DirectPayload",
            SubThreat::IllegalTarget { .. } => "IllegalTarget",
            SubThreat::RepeatedBase { .. } => "RepeatedBase",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Verdict {
    pub kind: ClassificationKind,
    pub sub: Option<SubThreat>,
    pub vs_positions: Vec<usize>,
    pub recovered_bytes: Vec<u8>,
}

fn decode_vs_run(input: &[u32], positions: &[usize]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut high: Option<u32> = None;
    for &p in positions {
        let n = match vs_to_nibble(input[p]) {
            Some(n) => n,
            None => continue,
        };
        match high {
            None => high = Some(n),
            Some(h) => {
                out.push(((h << 4) | n) as u8);
                high = None;
            }
        }
    }
    out
}

fn all_same_vs(input: &[u32], positions: &[usize]) -> bool {
    let cp0 = match positions.first() {
        Some(&p0) => input[p0],
        None => return true,
    };
    positions.iter().all(|&p| input[p] == cp0)
}

fn lossy_ascii(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|&b| {
            if (0x20..=0x7E).contains(&b) || b == 0x09 || b == 0x0A || b == 0x0D {
                b as char
            } else {
                '?'
            }
        })
        .collect()
}

pub fn detect(input: &[u32]) -> Verdict {
    let mut v = Verdict {
        kind: ClassificationKind::Clear,
        sub: None,
        vs_positions: Vec::new(),
        recovered_bytes: Vec::new(),
    };
    v.vs_positions = input
        .iter()
        .enumerate()
        .filter_map(|(i, &cp)| if is_variation_selector(cp) { Some(i) } else { None })
        .collect();

    if v.vs_positions.is_empty() {
        return v;
    }

    v.recovered_bytes = decode_vs_run(input, &v.vs_positions);

    // Single-VS exemption: if the entire VS run is exactly ONE
    // VS codepoint following a base, and that (base, VS) pair is
    // registered in StandardizedVariants or emoji-variation-
    // sequences, the input is a legitimate registered variation
    // (e.g. CJK Compatibility Ideograph + FE00, registered math
    // variant, or emoji-style/text-style selector).  Return Clear.
    if v.vs_positions.len() == 1 {
        let p = v.vs_positions[0];
        if p > 0 {
            let base = input[p - 1];
            let vs = input[p];
            if is_registered_variation_pair(base, vs) {
                // Legitimate variant — leave verdict Clear.
                return v;
            }
        }
    }

    v.kind = ClassificationKind::Hazard;

    if v.vs_positions.len() >= 4 && all_same_vs(input, &v.vs_positions) {
        let p0 = v.vs_positions[0];
        let base = if p0 == 0 { 0 } else { input[p0 - 1] };
        v.sub = Some(SubThreat::RepeatedBase {
            base_cp: base,
            vs_count: v.vs_positions.len(),
        });
    } else if !v.recovered_bytes.is_empty() {
        v.sub = Some(SubThreat::DirectPayload {
            decoded: lossy_ascii(&v.recovered_bytes),
        });
    } else {
        let p = v.vs_positions[0];
        let target = if p == 0 { 0 } else { input[p - 1] };
        v.sub = Some(SubThreat::IllegalTarget {
            target_cp: target,
            vs_cp: input[p],
        });
    }
    v
}
